### PR TITLE
some fixes to make it a real alternative to gnu msgfmt

### PR DESCRIPTION
**Pull request by xhebox**
1. automatically convert to UTF-8, if the encoding of file is unknown, then do nothing
2. v stands for verbose and V stands for version now
3. multiple plurals support
4. skip fuzzy marked strings
#9

### CODE REVIEW

**Comment by xhebox** [comment]
1. support converting to UTF-8:
Once charset is detected in get_type_and_start(), it's stored in info.charset. The real converting is in get_length_and_convert(). If iconv does not support the encoding then do nothing.

2. 
```
msgid ""
msgstr ""

msgid "xxx"
```
it's invalid that both msgid and msgstr are empty except the one at head. We need to ignore them. It's mainly done by dinvalid() in msgfmt.c.

3. multiple plurals:
we do not need to parse the complex phases, only nplurals is necessary(set to 2 by default). One check for plural is removed, and the following is added:
```
(x[0] == '[' && x[1] < info->nplurals && x[2] == ']' && (x += 3) && isspace(*x))))
```

since plurals are simply joined, i add a type to distinguish: pe_plural. Now we could copy the content but no counts for plurals.

4. Non c-format like sh-format is skipped(do we need to support more formats?)
5. fuzzy marked strings are skipped(do we need to add --use-fuzzy and -f options?)
6. action in actiontbl is more strict then before
7. v stands for verbos and V stands for version now

**Comment by xhebox** [comment]
1. plurals with sysdeps could be solved correctly:
```
msgid "%<PRIu32>"
msgid_plural "%<PRIu32> s"
msgstr[0] "%<PRIu32> a"
msgstr[1] "%<PRIu32> a"
```
2. autopoint is added to .gitignore

**Comment by rofl0r** [comment]
nice work! i didn't find time yet to dig into it, but i hope i'll do so in the next week.

File: src/msgfmt.c [hunk 1]
```diff
@@ -193,46 +202,93 @@ static void error(const char* msg) {
...
-		// kick the invalid out of counts
 		d->num[pe_msgid]--;
 		d->curr[pe_msgid]--;
 		d->num[pe_msgstr]--;
 		d->curr[pe_msgstr]--;
+		d->mslen1=d->mslen2=d->msc=0;
+		return;
+	}
+
+	if(d->msc && d->msc <= (info->nplurals-48)) {
```
Comment (rofl0r) [inline src/msgfmt.c]: what's that 48 there ?

File: src/msgfmt.c [hunk 2]
```diff
@@ -244,97 +300,66 @@ int process_line_callback(struct po_info* info, void* user) {
...
+						d->milen2 = l+1;
 					}
-					d->curr[pe_msgid]++;
 				} else if(info->type == pe_plural) {
-					if(pc != 2) {
-						memcpy(d->strbuffer[pe_msgid] + d->stroff[pe_msgid], sysdeps[i], l+1);
-						d->strlist[d->curr[pe_msgid]-1].str.len+=l+1;
-						d->stroff[pe_msgid]+=l+1;
+					if(i==0) {
+						assert(l+1 <= 4096);
```
Comment (rofl0r) [inline src/msgfmt.c]: in case that 4096 refers to a buffer size, it would be cleaner and more future-proof to use sizeof()

File: src/msgfmt.c (hunk 1 above)
Comment (xhebox) [inline src/msgfmt.c]: It's a ascii number, 0 is not 0 but 0x30(48)

File: src/msgfmt.c (hunk 2 above)
Comment (xhebox) [inline src/msgfmt.c]: how about now?

File: src/msgfmt.c (hunk 1 above)
Comment (rofl0r) [inline src/msgfmt.c]: oh, so  maybe `'0'` would make it clearer then.

File: src/poparser.c [hunk 3]
```diff
@@ -2,13 +2,14 @@
...
 #include <string.h>
+#include <iconv.h>
 #include "poparser.h"
 #include "StringEscape.h"
 
 #define streq(A, B) (!strcmp(A, B))
 #define strstarts(S, W) (memcmp(S, W, sizeof(W) - 1) ? NULL : (S + (sizeof(W) - 1)))
 
-static enum po_entry get_type_and_start(char* lp, char* end, size_t *stringstart) {
+static enum po_entry get_type_and_start(struct po_info *info, char* lp, char* end, size_t *stringstart) {
```
Comment (rofl0r) [inline src/poparser.c]: i think we can safely just pass char* charset here

File: src/poparser.h [hunk 4]
```diff
@@ -14,6 +14,7 @@ enum po_entry {
 struct po_info {
 	enum po_entry type;
 	char *text;
+	char charset[12];
```
Comment (rofl0r) [inline src/poparser.h]: probably makes more sense to put this into the po_parser struct

File: src/msgfmt.c [hunk 5]
```diff
@@ -289,16 +289,16 @@ int process(FILE *in, FILE *out) {
 	cb_for_qsort = &d;
 	qsort(d.strlist, d.num[pe_msgid], sizeof (struct strmap), strmap_comp);
 	unsigned i;
-	for(i = 0; i < d.num[0]; i++) {
```
Comment (rofl0r) [inline src/msgfmt.c]: please split into separate commit

File: src/msgfmt.c [hunk 6]
```diff
@@ -184,9 +193,50 @@ static void error(const char* msg) {
...
+	d->strlist[d->curr[pe_msgstr]].trans = &d->translist[d->curr[pe_msgstr]];
+
+	memcpy(d->strbuffer[pe_msgstr] + d->stroff[pe_msgstr], msgstr2, mslen2);
+	d->stroff[pe_msgstr]+=mslen2;
+	d->curr[pe_msgstr]++;
+
+	mslen1=mslen2=msc=0;
+}
+
+static inline void dinvalid(struct callbackdata *d, struct po_info *info) {
```
Comment (rofl0r) [inline src/msgfmt.c]: would be good to have a short description about the purpose of both function in the commit message

File: include/gettext.h [hunk 7]
```diff
@@ -0,0 +1,136 @@
+/* Convenience header for conditional use of GNU <libintl.h>.
```
Comment (rofl0r) [inline include/gettext.h]: please remove this commit, we can add the header at a later time when it will be needed

File: src/msgfmt.c [hunk 8]
```diff
@@ -323,34 +323,41 @@ int process_line_callback(struct po_info* info, void* user) {
 					writestr(d, info);
 					// just copy, it's written down when writemsg()
 					if(i==0) {
+						assert(l+1 <= sizeof(d->msgidbuf1));
```
Comment (rofl0r) [inline src/msgfmt.c]: please split into 2 commits
